src/strategies/saddle-finance/index.ts

```diff
@@ -0,0 +1,126 @@
+  'function beneficiary() external view returns (address)',
+  'function vestings(address) external view returns (bool isVerified, uint120 totalAmount, uint120 released)'
+];
+
+export async function strategy(
+  space,
+  network,
+  provider,
+  addresses,
+  options,
+  snapshot
+) {
+  const blockTag = typeof snapshot === 'number' ? snapshot : 'latest';
+
+  const userWalletBalanceResponse = await multicall(
+    network,
+    provider,
+    abi,
+    addresses.map((address: any) => [SDLTokenAddress, 'balanceOf', [address]]),
+    { blockTag }
+  );
+
+  const vestingAddrsBalanceRes = await multicall(
+    network,
+    provider,
+    abi,
+    vestingContractAddrs.map((vestingContractAddress: any) => [
+      SDLTokenAddress,
+      'balanceOf',
+      [vestingContractAddress]
+    ]),
+    { blockTag }
+  );
+
+  const beneficiaries = await multicall(
+    network,
+    provider,
+    abi,
+    vestingContractAddrs.map((vestingContractAddress: any) => [
+      vestingContractAddress,
+      'beneficiary'
+    ]),
+    { blockTag }
+  );
+
+  const retroAddrs = Object.keys(remappedMerkleData);
+
+  const userVestingsRes = await multicall(
+    network,
+    provider,
+    abi,
+    retroAddrs.map((retroAddr: any) => [
+      RetroRewardsContract,
+      'vestings',
+      [retroAddr]
+    ]),
+    { blockTag }
+  );
+
+  const retroUserBalances = {};
+  retroAddrs.forEach((addr, i) => {
+    const userVesting = userVestingsRes[i];
+    if (userVesting.isVerified) {
+      retroUserBalances[addr] = parseFloat(
+        formatUnits(
+          userVesting.totalAmount.sub(userVesting.released).toString(),
+          18
+        )
+      );
+    } else {
+      retroUserBalances[addr] = parseFloat(
+        formatUnits(remappedMerkleData[addr].amount, 18)
+      );
+    }
+  });
+
+  const mappedBeneficiariesToVestingContract = beneficiaries.reduce(
+    (acc, addr, i) => ({
+      ...acc,
+      [addr]: parseFloat(
+        formatUnits(vestingAddrsBalanceRes[i][0].toString(), 18)
+      )
+    }),
+    {}
+  );
+
+  const userWalletBalances = userWalletBalanceResponse.map((amount, i) => {
+    return [addresses[i], parseFloat(formatUnits(amount.toString(), 18))];
+  });
+
+  const userTotal = {};
+  // loop through user, investor/advisor/team-member, and airdrop wallets to calculate total.
+  userWalletBalances.forEach(([addr, amount]) => {
+    if (userTotal[addr]) userTotal[addr] += amount;
+    else userTotal[addr] = amount;
+  });
+  for (const [addr, amount] of Object.entries(retroUserBalances)) {
+    if (userTotal[addr]) userTotal[addr] += amount;
+    else userTotal[addr] = amount;
+  }
+  for (const [addr, amount] of Object.entries(
+    mappedBeneficiariesToVestingContract
+  )) {
+    if (userTotal[addr]) userTotal[addr] += amount;
+    else userTotal[addr] = amount;
+  }
+
+  const finalUserBalances = Object.fromEntries(
+    Object.entries(userTotal).filter(([addr]) => addresses.includes(addr))
```

Comment:
May be worth making `addresses` into a `Set` for faster lookups.

Reply:
Actually I guess it's faster just to do `addresses.map(addr => [addr, userTotal[addr] || 0])`

Reply:
It's a lot more readable that's for sure.